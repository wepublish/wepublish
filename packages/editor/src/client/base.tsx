import React, {ReactNode, useEffect, useState} from 'react'

import {Container, Sidebar, Sidenav, Nav, Navbar, Icon, Dropdown} from 'rsuite'

import {
  ArticleListRoute,
  useRoute,
  RouteType,
  PageListRoute,
  routeLink,
  AuthorListRoute,
  ImageListRoute,
  UserListRoute,
  UserRoleListRoute,
  PeerListRoute,
  TokenListRoute,
  MemberPlanListRoute,
  PaymentMethodListRoute,
  NavigationListRoute,
  LogoutRoute
} from './route'

import {useTranslation} from 'react-i18next'

export interface BaseProps {
  children?: ReactNode
}

const AVAILABLE_LANG = [
  {id: 'en', lang: 'en_US', name: 'English'},
  {id: 'fr', lang: 'fr_FR', name: 'FranÃ§ais'},
  {id: 'de', lang: 'de_CH', name: 'Deutsch'}
]

const iconStyles = {
  width: 56,
  height: 56,
  lineHeight: '56px',
  textAlign: 'center'
}
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const NavItemLink = routeLink(Nav.Item)
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const DropdownItemLink = routeLink(Dropdown.Item)

function useStickyState(defaultValue: string, key: string) {
  const [value, setValue] = useState(() => {
    const stickyValue = window.localStorage.getItem(key)
    return stickyValue !== null ? JSON.parse(stickyValue) : defaultValue
  })
  useEffect(() => {
    window.localStorage.setItem(key, JSON.stringify(value))
  }, [key, value])
  return [value, setValue]
}

export function Base({children}: BaseProps) {
  const {current} = useRoute()

  const {t, i18n} = useTranslation()

  const [isExpanded, setIsExpanded] = useState(true)

  const [uiLanguage, setUILanguage] = useStickyState(AVAILABLE_LANG[0].id, 'wepublish/language')

  useEffect(() => {
    i18n.changeLanguage(uiLanguage)
  }, [uiLanguage])

  return (
    <div style={{display: 'flex', height: '100vh', width: '100vw'}}>
      <Container>
        <Sidebar
          style={{display: 'flex', flexDirection: 'column'}}
          width={isExpanded ? 260 : 56}
          collapsible>
          <Sidenav
            expanded={isExpanded}
            defaultOpenKeys={['1']}
            appearance="subtle"
            style={{flex: '1 1 auto'}}>
            <Sidenav.Body>
              <Nav>
                <NavItemLink
                  icon={<Icon icon="file-text" />}
                  route={ArticleListRoute.create({})}
                  active={current?.type === RouteType.ArticleList}>
                  {t('navbar.articles')}
                </NavItemLink>

                <NavItemLink
                  icon={<Icon icon="file" />}
                  route={PageListRoute.create({})}
                  active={current?.type === RouteType.PageList}>
                  {t('navbar.pages')}
                </NavItemLink>

                <NavItemLink
                  icon={<Icon icon="people-group" />}
                  route={AuthorListRoute.create({})}
                  active={current?.type === RouteType.AuthorList}>
                  {t('navbar.authors')}
                </NavItemLink>

                <NavItemLink
                  icon={<Icon icon="camera-retro" />}
                  route={ImageListRoute.create({})}
                  active={current?.type === RouteType.ImageList}>
                  {t('navbar.imageLibrary')}
                </NavItemLink>

                <NavItemLink
                  icon={<Icon icon="bars" />}
                  route={NavigationListRoute.create({})}
                  active={current?.type === RouteType.NavigationList}>
                  {t('navbar.navigations')}
                </NavItemLink>

                <Dropdown
                  eventKey={'1'}
                  title={t('navbar.usersAndMembers')}
                  icon={<Icon icon="peoples" />}>
                  <DropdownItemLink
                    active={current?.type === RouteType.UserList}
                    icon={<Icon icon="user-circle" />}
                    route={UserListRoute.create({})}>
                    {t('navbar.users')}
                  </DropdownItemLink>

                  <DropdownItemLink
                    active={current?.type === RouteType.UserRoleList}
                    icon={<Icon icon="character-authorize" />}
                    route={UserRoleListRoute.create({})}>
                    {t('navbar.userRoles')}
                  </DropdownItemLink>

                  <DropdownItemLink
                    active={current?.type === RouteType.MemberPlanList}
                    icon={<Icon icon="id-card" />}
                    route={MemberPlanListRoute.create({})}>
                    {t('navbar.memberPlans')}
                  </DropdownItemLink>

                  <DropdownItemLink
                    active={current?.type === RouteType.PaymentMethodList}
                    icon={<Icon icon="money" />}
                    route={PaymentMethodListRoute.create({})}>
                    {t('navbar.paymentMethods')}
                  </DropdownItemLink>
                </Dropdown>

                <NavItemLink
                  icon={<Icon icon="share" />}
                  route={PeerListRoute.create({})}
                  active={current?.type === RouteType.PeerList}>
                  {t('navbar.peering')}
                </NavItemLink>

                <NavItemLink
                  icon={<Icon icon="key" />}
                  route={TokenListRoute.create({})}
                  active={current?.type === RouteType.TokenList}>
                  {t('navbar.tokens')}
                </NavItemLink>
              </Nav>
            </Sidenav.Body>
          </Sidenav>
          <Navbar appearance="subtle" className="nav-toggle">
            <Navbar.Body>
              <Nav>
                <Dropdown
                  placement="topStart"
                  trigger="click"
                  renderTitle={children => {
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore
                    return <Icon style={iconStyles} icon="cog" />
                  }}>
                  <DropdownItemLink route={LogoutRoute.create({})}>
                    {t('navbar.logout')}
                  </DropdownItemLink>
                </Dropdown>
              </Nav>
              <Nav>
                <Dropdown
                  placement="topStart"
                  trigger="click"
                  renderTitle={() => (
                    <Icon
                      icon="globe"
                      style={{
                        width: 56,
                        height: 56,
                        lineHeight: '56px',
                        textAlign: 'center'
                      }}
                    />
                  )}>
                  {AVAILABLE_LANG.map(lang => (
                    <Dropdown.Item
                      key={lang.id}
                      onSelect={() => setUILanguage(lang.id)}
                      active={lang.id === uiLanguage}>
                      {lang.name}
                    </Dropdown.Item>
                  ))}
                </Dropdown>
              </Nav>

              <Nav pullRight>
                <Nav.Item
                  onClick={() => setIsExpanded(!isExpanded)}
                  style={{width: 56, textAlign: 'center'}}>
                  <Icon icon={isExpanded ? 'angle-left' : 'angle-right'} />
                </Nav.Item>
              </Nav>
            </Navbar.Body>
          </Navbar>
        </Sidebar>
        <Container
          style={{
            paddingTop: '60px',
            paddingBottom: '60px',
            paddingLeft: '40px',
            paddingRight: '40px'
          }}>
          {children}
        </Container>
      </Container>
    </div>
  )
}
