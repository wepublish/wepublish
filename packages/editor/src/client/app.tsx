import React from 'react'
import {hot} from 'react-hot-loader/root'

import 'rsuite/lib/styles/index.less'

import {useRoute, RouteType, Route} from './route'

import {Login} from './login'
import {Base} from './base'
import {ArticleList} from './routes/articleList'

import {ArticleEditor} from './routes/articleEditor'
import {ImageList} from './routes/imageList'
import {PageList} from './routes/pageList'
import {PageEditor} from './routes/pageEditor'
import {AuthorList} from './routes/authorList'
import {PeerList} from './routes/peerList'
import {TokenList} from './routes/tokenList'
import {UserList} from './routes/userList'
import {CommentList} from './routes/commentList'
import {UserRoleList} from './routes/userRoleList'
import {MemberPlanList} from './routes/memberPlanList'
import {PaymentMethodList} from './routes/paymentMethodList'
import {NavigationList} from './routes/navigationList'

export function contentForRoute(route: Route) {
  switch (route.type) {
    case RouteType.Login:
      return <Login />

    case RouteType.PeerList:
    case RouteType.PeerProfileEdit:
    case RouteType.PeerCreate:
    case RouteType.PeerEdit:
      return <PeerList />

    case RouteType.TokenList:
    case RouteType.TokenGenerate:
      return <TokenList />

    case RouteType.Index:
    case RouteType.ArticleList:
      return <ArticleList />

    case RouteType.CommentList:
      return <CommentList />

    case RouteType.PageList:
      return <PageList />

    case RouteType.ImageList:
    case RouteType.ImageUpload:
    case RouteType.ImageEdit:
      return <ImageList />

    case RouteType.AuthorList:
    case RouteType.AuthorCreate:
    case RouteType.AuthorEdit:
      return <AuthorList />

    case RouteType.UserList:
    case RouteType.UserCreate:
    case RouteType.UserEdit:
      return <UserList />

    case RouteType.UserRoleList:
    case RouteType.UserRoleCreate:
    case RouteType.UserRoleEdit:
      return <UserRoleList />

    case RouteType.MemberPlanList:
    case RouteType.MemberPlanCreate:
    case RouteType.MemberPlanEdit:
      return <MemberPlanList />

    case RouteType.PaymentMethodList:
    case RouteType.PaymentMethodCreate:
    case RouteType.PaymentMethodEdit:
      return <PaymentMethodList />

    case RouteType.NavigationList:
    case RouteType.NavigationCreate:
    case RouteType.NavigationEdit:
      return <NavigationList />

    case RouteType.NotFound:
      return <ArticleList />
  }

  return null
}

export function App() {
  const {current} = useRoute()

  if (current) {
    switch (current.type) {
      case RouteType.Login:
        return <Login />

      case RouteType.ArticleCreate:
      case RouteType.ArticleEdit:
        return (
          <ArticleEditor
            id={current.type === RouteType.ArticleEdit ? current.params.id : undefined}
          />
        )

      case RouteType.PageCreate:
      case RouteType.PageEdit:
        return (
          <PageEditor id={current.type === RouteType.PageEdit ? current.params.id : undefined} />
        )

      default:
        return <Base>{contentForRoute(current)}</Base>
    }
  }

  return null
}

export const HotApp = hot(App)
