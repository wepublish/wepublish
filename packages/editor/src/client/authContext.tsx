import React from 'react'
import {createContext, Dispatch, useReducer, ReactNode} from 'react'

export interface AuthContextState {
  readonly session?: {
    readonly username: string
    readonly token: string
  }
}

export const AuthContext = createContext<AuthContextState>({})

export enum AuthDispatchActionType {
  Login = 'login',
  Logout = 'logout'
}

export interface AuthDispatchLoginAction {
  readonly type: AuthDispatchActionType.Login
  readonly username: string
  readonly token: string
}

export interface AuthDispatchLogoutAction {
  readonly type: AuthDispatchActionType.Logout
}

export type AuthDispatchAction = AuthDispatchLoginAction | AuthDispatchLogoutAction

export const AuthDispatchContext = createContext<Dispatch<AuthDispatchAction>>(() => {
  throw new Error('No AuthProvider found in component tree.')
})

export function authReducer(
  _prevState: AuthContextState,
  action: AuthDispatchAction
): AuthContextState {
  switch (action.type) {
    case AuthDispatchActionType.Login:
      return {session: {username: action.username, token: action.token}}

    case AuthDispatchActionType.Logout:
      return {}
  }
}

export interface AuthProviderProps {
  readonly children?: ReactNode
}

export function AuthProvider({children}: AuthProviderProps) {
  const [state, dispatch] = useReducer(authReducer, {})

  return (
    <AuthDispatchContext.Provider value={dispatch}>
      <AuthContext.Provider value={state}>{children}</AuthContext.Provider>
    </AuthDispatchContext.Provider>
  )
}
