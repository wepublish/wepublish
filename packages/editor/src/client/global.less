.rs-table {
  .highlighted-row,
  .highlighted-row &-cell-group &-cell {
    background-color: #f2faff;
  }

  &-row.approved:not(&-row-header),
  &-row.approved &-cell-group,
  &-row.approved &-cell {
    background: rgb(225, 248, 222);
  }

  &-row.pending-user:not(&-row-header),
  &-row.pending-user &-cell-group,
  &-row.pending-user &-cell {
    background: rgb(248, 239, 222);
  }

  &-row.pending-approval:not(&-row-header),
  &-row.pending-approval &-cell-group,
  &-row.pending-approval &-cell {
    background: rgb(248, 222, 242);
  }

  &-row.rejected:not(&-row-header),
  &-row.rejected &-cell-group,
  &-row.rejected &-cell {
    background: rgb(83, 85, 83);
    color: white;
    text-decoration: line-through;
  }

  &-hover &-row.rejected:not(&-row-header):hover,
  &-hover &-row.rejected:hover &-cell-group,
  &-hover &-row.rejected:hover &-cell {
    color: rgb(83, 85, 83);
    text-decoration: line-through;
  }
}
