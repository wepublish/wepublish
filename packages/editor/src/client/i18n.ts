import i18n from 'i18next'
import {initReactI18next} from 'react-i18next'
import LanguageDetector from 'i18next-browser-languagedetector'

import en from './locales/en.json'
import fr from './locales/fr.json'
import de from './locales/de.json'

i18n.use(LanguageDetector).use(initReactI18next).init({
  fallbackLng: 'en',
  debug: false,
  resources: {
    en,
    fr,
    de
  }
})

export default i18n
