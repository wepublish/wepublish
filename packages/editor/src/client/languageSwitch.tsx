/* eslint-disable i18next/no-literal-string */

import React, {useEffect, useState} from 'react'
import {
  Select,
  Dialog,
  Panel,
  PanelSection,
  PanelHeader,
  NavigationButton,
  MenuButton
} from '@karma.run/ui'
import {cssRule, useStyle} from '@karma.run/react'
import {MaterialIconClose, MaterialIconLanguage} from '@karma.run/icons'
import {useTranslation} from 'react-i18next'

export function pxToRem(px: number): string {
  return `${px / 10}rem`
}

export function useStickyState(
  defaultValue: {id: string; lang: string; name: string},
  key: string
) {
  const [value, setValue] = useState(() => {
    const stickyValue = window.localStorage.getItem(key)
    return stickyValue !== null ? JSON.parse(stickyValue) : defaultValue
  })
  useEffect(() => {
    window.localStorage.setItem(key, JSON.stringify(value))
  }, [key, value])
  return [value, setValue]
}

let AVAILABLE_LANG: {id: string; lang: string; name: string}[] = []

export function LanguageSelector() {
  AVAILABLE_LANG = [
    {id: 'en', lang: 'en_US', name: 'English'},
    {id: 'fr', lang: 'fr_FR', name: 'Fran√ßais'},
    {id: 'de', lang: 'de_CH', name: 'Deutsch'}
  ]

  const [uiLanguage, setUILanguage] = useStickyState(AVAILABLE_LANG[0], 'localStorageValues')

  const {i18n} = useTranslation()

  const css = useStyle()
  const languageSelector = cssRule({
    marginBottom: pxToRem(62.5)
  })

  return (
    <div className={css(languageSelector)}>
      <Select
        options={AVAILABLE_LANG}
        value={AVAILABLE_LANG.find(lang => lang.id === uiLanguage.id)}
        renderListItem={uiLanguage => uiLanguage.name}
        onChange={values => {
          if (values?.name) {
            setUILanguage(values)
            i18n.changeLanguage(values.id)
          }
        }}
      />
    </div>
  )
}

export function LanguageSwitch() {
  const [isLanguageSwitchDialogOpen, setIsLanguageSwitchDialogOpen] = useState(false)
  const {t} = useTranslation()

  return (
    <>
      <MenuButton
        icon={MaterialIconLanguage}
        label={t('navbar.languageSwitch.currentLanguage')}
        onClick={() => setIsLanguageSwitchDialogOpen(true)}
      />
      <Dialog
        open={isLanguageSwitchDialogOpen}
        onClose={() => setIsLanguageSwitchDialogOpen(false)}
        width={400}
        closeOnBackgroundClick>
        {() => (
          <Panel>
            <PanelHeader
              title={t('navbar.languageSwitch.selectNewLanguage')}
              leftChildren={
                <NavigationButton
                  icon={MaterialIconClose}
                  label={t('navbar.languageSwitch.close')}
                  onClick={() => setIsLanguageSwitchDialogOpen(false)}
                />
              }
            />
            <PanelSection>
              <LanguageSelector />
            </PanelSection>
          </Panel>
        )}
      </Dialog>
    </>
  )
}
