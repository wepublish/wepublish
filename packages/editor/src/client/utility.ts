import nanoid from 'nanoid'
import {useCallback, useEffect, useMemo, useRef, useState} from 'react'
import {DocumentNode, OperationDefinitionNode} from 'graphql'
import {PaymentPeriodicity, SortOrder} from './api'

export enum LocalStorageKey {
  SessionToken = 'sessionToken'
}

export function generateID(): string {
  return nanoid()
}

// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
    .replace(/[ÇĆĈČ]/gi, 'c')
    .replace(/[ÐĎĐÞ]/gi, 'd')
    .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
    .replace(/[ĜĞĢǴ]/gi, 'g')
    .replace(/[ĤḦ]/gi, 'h')
    .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
    .replace(/[Ĵ]/gi, 'j')
    .replace(/[Ĳ]/gi, 'ij')
    .replace(/[Ķ]/gi, 'k')
    .replace(/[ĹĻĽŁ]/gi, 'l')
    .replace(/[Ḿ]/gi, 'm')
    .replace(/[ÑŃŅŇ]/gi, 'n')
    .replace(/[ÒÓÔÕÖØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
    .replace(/[Œ]/gi, 'oe')
    .replace(/[ṕ]/gi, 'p')
    .replace(/[ŔŖŘ]/gi, 'r')
    .replace(/[ßŚŜŞŠ]/gi, 's')
    .replace(/[ŢŤ]/gi, 't')
    .replace(/[ÙÚÛÜŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
    .replace(/[ẂŴẀẄ]/gi, 'w')
    .replace(/[ẍ]/gi, 'x')
    .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
    .replace(/[ŹŻŽ]/gi, 'z')
    .replace(/[·/_,:;\\']/gi, '-')
    .replace(/\s+/g, '-')
    .replace(/[^\w\-]+/g, '') //eslint-disable-line
    .replace(/--+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '')
}

// https://gist.github.com/WebReflection/6076a40777b65c397b2b9b97247520f0
export function dateTimeLocalString(date: Date) {
  function prefix(i: number) {
    return (i < 10 ? '0' : '') + i
  }

  const year = date.getFullYear()
  const month = prefix(date.getMonth() + 1)
  const day = prefix(date.getDate())
  const hours = prefix(date.getHours())
  const minutes = prefix(date.getMinutes())
  const seconds = prefix(date.getSeconds())

  return `${year}-${month}-${day}T${hours}:${minutes}:${seconds}`
}

export function useScript(src: string, checkIfLoaded: () => boolean, crossOrigin = false) {
  if (typeof window !== 'object')
    return {
      isLoaded: false,
      isLoading: false,
      load: () => {
        /* do nothing */
      }
    }

  const scriptRef = useRef<HTMLScriptElement | null>(null)

  const [isLoading, setLoading] = useState(false)
  const [isLoaded, setLoaded] = useState(() => checkIfLoaded())

  useEffect(() => {
    if (isLoading && !isLoaded && !scriptRef.current) {
      const script = document.createElement('script')

      script.src = src
      script.async = true
      script.defer = true
      script.onload = () => setLoaded(true)
      script.crossOrigin = crossOrigin ? 'anonymous' : null

      document.head.appendChild(script)
      scriptRef.current = script
    }
  }, [isLoading])

  const load = useCallback(() => {
    setLoading(true)
  }, [])

  return useMemo(
    () => ({
      isLoading,
      isLoaded,
      load
    }),
    [isLoading, isLoaded, load]
  )
}

export function getOperationNameFromDocument(node: DocumentNode) {
  const firstOperation = node.definitions.find(
    node => node.kind === 'OperationDefinition'
  ) as OperationDefinitionNode

  if (!firstOperation?.name?.value) throw new Error("Coulnd't find operation name.")
  return firstOperation.name.value
}

export function transformCssStringToObject(styleCustom: string): Record<string, unknown> {
  const styleRules = styleCustom.split(';')
  if (styleRules.length === 0) return {}
  return styleRules.reduce((previousValue: Record<string, unknown>, currentValue: string) => {
    const [key, value] = currentValue.split(':')
    if (key && value) {
      return Object.assign(previousValue, {[key.trim()]: value.trim()})
    }
    return previousValue
  }, {})
}

export type SortType = 'asc' | 'desc' | null
export function mapTableSortTypeToGraphQLSortOrder(sortType: SortType): SortOrder | null {
  switch (sortType) {
    case 'desc':
      return SortOrder.Descending
    case 'asc':
      return SortOrder.Ascending
    default:
      return null
  }
}

export const DEFAULT_TABLE_PAGE_SIZES = [
  {
    value: 10,
    label: 10
  },
  {
    value: 20,
    label: 20
  },
  {
    value: 50,
    label: 50
  },
  {
    value: 100,
    label: 100
  }
]

export const ALL_PAYMENT_PERIODICITIES: PaymentPeriodicity[] = [
  PaymentPeriodicity.Monthly,
  PaymentPeriodicity.Quarterly,
  PaymentPeriodicity.Biannual,
  PaymentPeriodicity.Yearly
]
