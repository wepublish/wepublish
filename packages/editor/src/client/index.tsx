import 'core-js/stable'
import 'regenerator-runtime/runtime'

import React from 'react'
import ReactDOM from 'react-dom'
import {ApolloProvider, ApolloClient, ApolloLink, InMemoryCache} from '@apollo/client'
import {createUploadLink} from 'apollo-upload-client'
import {onError} from '@apollo/client/link/error'

import './i18n'

import {ElementID} from '../shared/elementID'
import {ClientSettings} from '../shared/types'
import {RouteProvider} from './route'
import {AuthProvider} from './authContext'
import {LocalStorageKey} from './utility'
import {TwitterProvider} from './blocks/embeds/twitter'
import {InstagramProvider} from './blocks/embeds/instagram'
import {FacebookProvider} from './blocks/embeds/facebook'
import {HotApp} from './app'

// See: https://www.apollographql.com/docs/react/data/fragments/#fragments-on-unions-and-interfaces
export async function fetchIntrospectionQueryResultData(url: string) {
  const response = await fetch(url, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({
      variables: {},
      query: `
      {
        __schema {
          types {
            kind
            name
            possibleTypes {
              name
            }
          }
        }
      }
    `
    })
  })

  const result = await response.json()

  const possibleTypes: any = {}

  result.data.__schema.types.forEach((supertype: any) => {
    if (supertype.possibleTypes) {
      possibleTypes[supertype.name] = supertype.possibleTypes.map((subtype: any) => subtype.name)
    }
  })

  return possibleTypes
}

const onDOMContentLoaded = async () => {
  const {apiURL}: ClientSettings = JSON.parse(
    document.getElementById(ElementID.Settings)!.textContent!
  )

  const adminAPIURL = `${apiURL}/admin`

  const authLink = new ApolloLink((operation, forward) => {
    const token = localStorage.getItem(LocalStorageKey.SessionToken)
    const context = operation.getContext()

    operation.setContext({
      headers: {
        authorization: token ? `Bearer ${token}` : '',
        ...context.headers
      },
      ...context
    })

    return forward(operation)
  })

  const authErrorLink = onError(({graphQLErrors, /* networkError, */ operation, forward}) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({/* message, locations, path, */ extensions}) => {
        // console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`)
        if (extensions?.code === 'UNAUTHENTICATED') {
          window.location.pathname = '/logout'
        }
      })
      // if (networkError) console.log(`[Network error]: ${networkError}`)
    }
    return forward(operation)
  })

  const client = new ApolloClient({
    link: authLink.concat(authErrorLink).concat(createUploadLink({uri: adminAPIURL})),
    cache: new InMemoryCache({
      possibleTypes: await fetchIntrospectionQueryResultData(adminAPIURL)
    })
  })

  window.addEventListener('dragover', e => e.preventDefault())
  window.addEventListener('drop', e => e.preventDefault())

  ReactDOM.render(
    <ApolloProvider client={client}>
      <AuthProvider>
        <RouteProvider>
          <FacebookProvider sdkLanguage={'en_US'}>
            <InstagramProvider>
              <TwitterProvider>
                <HotApp />
              </TwitterProvider>
            </InstagramProvider>
          </FacebookProvider>
        </RouteProvider>
      </AuthProvider>
    </ApolloProvider>,
    document.getElementById(ElementID.ReactRoot)
  )
}

if (document.readyState !== 'loading') {
  onDOMContentLoaded()
} else {
  document.addEventListener('DOMContentLoaded', onDOMContentLoaded)
}
