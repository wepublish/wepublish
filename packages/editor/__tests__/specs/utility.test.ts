import {slugify} from '../../src/client/utility'
describe('slugify', () => {
  test.each([
    'text with spaces',
    'text with UPPERCASE',
    '    text starting with spaces',
    'text with special characters like ·/_,:;\\',
    'text with hyphenized-words',
    'text with double -- dashes ',
    'text with multiple ------------- dashes',
    '- text starting with a dash',
    'text ending with a dash -',
    'umlaute like ä ö and ü',
    'special a like ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ',
    'special c like ÇĆĈČ',
    'special d like ÐĎĐÞ',
    'special e like ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ',
    'special g like ĜĞĢǴ',
    'special h like ĤḦ',
    'special i like ÌÍÎÏĨĪĮİỈỊ',
    'special j like Ĵ',
    'special ij like Ĳ',
    'special k like Ķ',
    'special l like ĹĻĽŁ',
    'special m like Ḿ',
    'special n like ÑŃŅŇ',
    'special o like ÒÓÔÕØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ',
    'special oe like Œ',
    'special p like ṕ',
    'special r like ŔŖŘ',
    'special s like ŚŜŞŠ',
    'special ss like ß',
    'special t like ŢŤ',
    'special u like ÙÚÛŨŪŬŮŰŲỤỦỨỪỬỮỰƯ',
    'special w like ẂŴẀẄ',
    'special x like ẍ',
    'special y like ÝŶŸỲỴỶỸ',
    'special z like ŹŻŽ'
  ])('should transform %s correctly', text => {
    const result = slugify(text)
    expect(result).toMatchSnapshot()
  })
})
