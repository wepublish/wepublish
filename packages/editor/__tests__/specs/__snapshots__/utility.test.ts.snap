// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slugify should transform     text starting with spaces correctly 1`] = `"text-starting-with-spaces"`;

exports[`slugify should transform - text starting with a dash correctly 1`] = `"text-starting-with-a-dash"`;

exports[`slugify should transform special a like ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ correctly 1`] = `"special-a-like-aaaaaeaaaaaaaaaaaaaaaaa"`;

exports[`slugify should transform special c like ÇĆĈČ correctly 1`] = `"special-c-like-cccc"`;

exports[`slugify should transform special d like ÐĎĐÞ correctly 1`] = `"special-d-like-dddd"`;

exports[`slugify should transform special e like ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ correctly 1`] = `"special-e-like-eeeeeeeeeeeeeeeee"`;

exports[`slugify should transform special g like ĜĞĢǴ correctly 1`] = `"special-g-like-gggg"`;

exports[`slugify should transform special h like ĤḦ correctly 1`] = `"special-h-like-hh"`;

exports[`slugify should transform special i like ÌÍÎÏĨĪĮİỈỊ correctly 1`] = `"special-i-like-iiiiiiiiii"`;

exports[`slugify should transform special ij like Ĳ correctly 1`] = `"special-ij-like-ij"`;

exports[`slugify should transform special j like Ĵ correctly 1`] = `"special-j-like-j"`;

exports[`slugify should transform special k like Ķ correctly 1`] = `"special-k-like-k"`;

exports[`slugify should transform special l like ĹĻĽŁ correctly 1`] = `"special-l-like-llll"`;

exports[`slugify should transform special m like Ḿ correctly 1`] = `"special-m-like-m"`;

exports[`slugify should transform special n like ÑŃŅŇ correctly 1`] = `"special-n-like-nnnn"`;

exports[`slugify should transform special o like ÒÓÔÕØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ correctly 1`] = `"special-o-like-ooooooooooooooooooooooo"`;

exports[`slugify should transform special oe like Œ correctly 1`] = `"special-oe-like-oe"`;

exports[`slugify should transform special p like ṕ correctly 1`] = `"special-p-like-p"`;

exports[`slugify should transform special r like ŔŖŘ correctly 1`] = `"special-r-like-rrr"`;

exports[`slugify should transform special s like ŚŜŞŠ correctly 1`] = `"special-s-like-ssss"`;

exports[`slugify should transform special ss like ß correctly 1`] = `"special-ss-like-ss"`;

exports[`slugify should transform special t like ŢŤ correctly 1`] = `"special-t-like-tt"`;

exports[`slugify should transform special u like ÙÚÛŨŪŬŮŰŲỤỦỨỪỬỮỰƯ correctly 1`] = `"special-u-like-uuuuuuuuuuuuuuuuu"`;

exports[`slugify should transform special w like ẂŴẀẄ correctly 1`] = `"special-w-like-wwww"`;

exports[`slugify should transform special x like ẍ correctly 1`] = `"special-x-like-x"`;

exports[`slugify should transform special y like ÝŶŸỲỴỶỸ correctly 1`] = `"special-y-like-yyyyyyy"`;

exports[`slugify should transform special z like ŹŻŽ correctly 1`] = `"special-z-like-zzz"`;

exports[`slugify should transform text ending with a dash - correctly 1`] = `"text-ending-with-a-dash"`;

exports[`slugify should transform text with UPPERCASE correctly 1`] = `"text-with-uppercase"`;

exports[`slugify should transform text with double -- dashes  correctly 1`] = `"text-with-double-dashes"`;

exports[`slugify should transform text with hyphenized-words correctly 1`] = `"text-with-hyphenized-words"`;

exports[`slugify should transform text with multiple ------------- dashes correctly 1`] = `"text-with-multiple-dashes"`;

exports[`slugify should transform text with spaces correctly 1`] = `"text-with-spaces"`;

exports[`slugify should transform text with special characters like ·/_,:;\\ correctly 1`] = `"text-with-special-characters-like"`;

exports[`slugify should transform umlaute like ä ö and ü correctly 1`] = `"umlaute-like-ae-oe-and-ue"`;
